Initial design doc for AWS GPU support

therc opened the pull request:
This is really v0

cc @erictune @davidopp @ravigadde @Hui-Zhi @Random-Liu

Commit:
Initial design doc for AWS GPU support

hongchaodeng commented on docs/proposals/gpu-support.md:
```diff
@@ -0,0 +1,263 @@
+in, requiring two devices per pod, but it can't schedule anywhere, because all
+we can find, at most, is one unused device per node.
+
+For v0, we'll use [a scheduler extender](../design/scheduler_extender.md). An
```
Why extender? Why not use node affinity?

mqliang replied:
Same question.

therc replied:
Node affinity only gives you a node with the right kind of resources, right? It doesn't track how much of that is allocated to others. With pod affinity, things could get a bit better, but then the user would have to plan carefully, especially if they have more than two replication controllers asking for GPUs. I'm trying to support resource sharing.

hongchaodeng commented on docs/proposals/gpu-support.md:
```diff
@@ -0,0 +1,263 @@
+we can find, at most, is one unused device per node.
+
+For v0, we'll use [a scheduler extender](../design/scheduler_extender.md). An
+extender runs outside of `kube-scheduler` and is consulted before picking a
```
No need to explain extender design. Link is good enough.

therc replied:
There are important details that are not obvious from the extender design document, e.g. the fact that extenders are run last and that they themselves are run in order.

hongchaodeng commented on docs/proposals/gpu-support.md:
```diff
@@ -0,0 +1,263 @@
+### The runtime
+
+Once we know where to run the container, it's time to set up its environment. We
+rely on the [nvidia-docker plugin](https://github.com/NVIDIA/nvidia-docker),
```
This sounds too specific.
Can you rephrase it to be more general for design purpose? It reads like taking notes for your company's workflow.

A good layout would be:
- Using GPU in containers in general
- nvidia-docker plugin

therc replied:
Done, but there's still a lot more in the nvidia-docker section than in the general one.

I'd love to have more generic infrastructure from the beginning, but in reality I can't find anyone trying to use non-NVIDIA devices in containers. The closest is this, which hasn't been touched in a year and only deals with creating OpenCL images:

https://github.com/chesterkuo/docker-gpgpu

I doubt it's a coincidence that both AWS and Azure support NVIDIA only. Same with Mesos. I can probably find a contact at Intel and inquire. Not sure about AMD. It looks like only NVIDIA has libraries (in Go!) and plugins to abstract (a little) the hardware and the drivers/libraries controlling it, to expose all that to a container engine.

hongchaodeng commented on docs/proposals/gpu-support.md:
```diff
@@ -0,0 +1,263 @@
+map devices into the containers and manage the CUDA volumes. As shown above,
+volumes are named after the hardware driver version.
+
+### Timeline
```
Please make it steps, not timelines. In each step, instead of talking about features, please lay out details of implementation steps.

therc replied:
Done. Do the steps need even more details?

hongchaodeng commented on docs/proposals/gpu-support.md:
What sample output? Can you clarify?

hongchaodeng commented on docs/proposals/gpu-support.md:
What does this command do?

hongchaodeng commented on docs/proposals/gpu-support.md:
Sorry, let me clarify with example. See [this](https://github.com/kubernetes/kubernetes/blob/7b3d1b370f80b3087a36f8e383de65edbb048f80/docs/design/podaffinity.md#implementation-plan).

The step you wrote is correct now. But please change "timeline" to "step" style without specific release target. When people are looking into the repo and reading proposals, they want to understand how it works. They can/should track release versions in other ways.

hongchaodeng commented on docs/proposals/gpu-support.md:
Can you clarify what "resource plugin" is/are?

therc commented on docs/proposals/gpu-support.md:
The plugin runs as a standalone daemon, listening to HTTP requests. One of the endpoints returns flags to add to Docker command lines. By default, it assumes you want to run on device 0, but you can change the URL to provide a list of device numbers instead.

hongchaodeng commented on docs/proposals/gpu-support.md:
Can you provide more implementation details for how to add plugin and how to call plugins?

hongchaodeng commented on docs/proposals/gpu-support.md:
Sorry for bothering.

Please talk about implementation details only. Track timeline in other ways.
It sounds like goals. If so, move it to "Objective" section.

hongchaodeng commented on docs/proposals/gpu-support.md:
I'm all supportive of what you wrote in V0 here.
But for the other Vs, it really just look like individual timeline that should be tracked in other ways.
You can add it to "Future Plan", "Objective", or submit a issue discussing your timeline.
But please write how to achieve things in proposal. Thanks!

hongchaodeng commented:
@Random-Liu @davidopp 
Can you add "e2e-not-required" for this?

therc commented on docs/proposals/gpu-support.md:
I thought I would revisit the document later, which is why only V0 is detailed. I removed release targets and changed the title to Implementation Plan.

hongchaodeng commented on docs/proposals/gpu-support.md:
Great. Thanks!

therc commented on docs/proposals/gpu-support.md:
Added a description of a new command line flag and the HTTP endpoints (assuming we stick with HTTP, which is up for debate).

davidopp commented:
I don't think we should use scheduler extender for this. I think scheduler extender is fine for custom extensions that individual users want to build, but we shouldn't use it for core resources. (Even for prototyping I don't think we should use it.)

I think the minimum requirement for the first version is to treat GPU as a first-class integer resource. That means not using the node label scheme you described, and instead implementing all of the things in your V1 section except "report resource usage" -- in particular, Kubelet should discover the number of GPUs on the machine and report it in node capacity, and the scheduler predicate that checks for sufficient resources should also check for sufficient GPU. I think it's fine for us to be very NVIDIA-specific; we can call the resource "nvidiaGpu" or something. My only concern is if we want to make it a fractional resource later. (I agree the first version should not support that.) In that case perhaps it should be something like nvidiaMilliGpu and validation should just reject values that are not multiples of 1000.

I read the nvidia-docker documentations and was trying to reconcile it with your proposal. They seem to encourage people to start the containers using nvidia-docker. IIUC your proposal is for Kubelet to get parameters using nvidia-docker-plugin, and then Kubelet can somehow start the container using that information, without using nvidia-docker? Is this what you were referring to when you wrote "However we obtain the data, we'll need to change Kubelet's dockertools to also map devices into the containers and manage the CUDA volumes"? Can you add more detail about how this works?  

With these comments in mind, it would be great if you could write an end-to-end list of steps of everything that happens from when the user creates their docker container (which IIUC requires them to know something about the nodes' hardware and/or software, based on what you wrote in the Runtime section), until it is running on a node, covering everything that is involved, including details of Kubelet and nvidia-docker-plugin. Assume "external binary" approach for running nvidia-docker-plugin. And as I mentioned earlier, I think it's fine to be NVIDIA-specific. That will be challenging enough without trying to generalize. :)

Thanks a lot for working on this! Although we hadn't planned to incorporate GPU so soon, we are happy to review your proposals and implementation (within the limitations of our review bandwidth).

cc/ @vishh @dchen1107 @yujuhong @erictune

therc commented:
Re: fractional resources. The more common use case for that (which happens to be ours, too...) is memory. GPUs come with more and more of it, but some workloads don't grow organically as fast. Of course it's up to the user if they want to trade utilization for increased context switching times. Some workloads wait for data to come from the CPU, disk or even the network.

We can't use nvidia-docker, even if NVIDIA recommends in their docs. That's a thin wrapper over the docker command line. We use a Go client library (third-party currently, the official one soon). So we need to obtain the information about devices and volumes from nvidia-docker-plugin (or link directly against its libraries) and pass that to the Docker client. On paper, we could run Docker/nvidia-docker for those special pods that need a GPU, but I suspect it would break so many things — or we would have to reimplement them for these frankenpods. In one of the issues, NVIDIA even encouraged us to use the library instead of the plugin.

I'll add the "turn-by-turn" description of what happens from pod creation to the moment the application binary starts (as @erictune mentioned in the parallel doc review, I should also mention explicitly what's up to the administrator and what's up to us).

davidopp commented:
> The more common use case for that (which happens to be ours, too...) is memory. 

Just to be clear, you're not suggesting we treat GPU processing and memory separately, right? Just that we treat GPU as fractional?

> We use a Go client library

Please include an explanation of how this works when you write the step-by-step instructions. It wasn't clear to me where the Go client library comes in (is it linked into Kubelet and used to start the GPU-using container?).

hongchaodeng commented on docs/proposals/gpu-support.md:
Please remove the first line "An obvious improvement...". Or put it to previous paragraph.
It's really confusing here

hongchaodeng commented on docs/proposals/gpu-support.md:
What do yo mean by "linked to kubelet"? How can containers access the information with such?

hongchaodeng commented:
@therc 
That would be great if you can provide detailed description of your use case. One thing I didn't catch is why it needs to link something into kubelet? From what I understand, you can install GPU drivers in OS. The mapping comes to container image level.

therc commented on docs/proposals/gpu-support.md:
It might sound counter-intuitive, but the containers themselves don't need information about which device to use or the actual Docker volume name. The kubelet needs all that, in order to set up the environment in the fairly homogeneous fashion that the applications expect. In more detail:

The NVIDIA client libraries embedded in user applications know to probe for all /dev/nvidia0, /dev/nvidia1 device nodes, etc. when they start (they don't even need to start from 0). The kubelet, on the other hand, needs to know which devices have to be exposed from the host into the container. For v0, we always map /dev/nvidia0 into the container, but in the future we might for example expose nvidia1 and nvidia3 when a pod that needs two GPUs lands on a machine that has four cards. The application will just detect and use those two. Same for the Docker volume: the application just sees NVIDIA's binaries and shared libraries under /bin and /lib, but the kubelet (with Docker's and nvidia-docker-plugin's help) needs to set that up correctly. The question is whether the kubelet should query for this information by sending an HTTP request to a daemon that runs on the same machine or if we just embed the daemon, by compiling and linking NVIDIA's C/Go library into the kubelet itself.

therc commented:
@hongchaodeng the linking is to have NVIDIA's Docker abstraction library (not the kernel driver, not the application shared libraries) built into the same binary as the kubelet, rather than in a daemon on the same machine that talks HTTP. Building it in makes a lot of failure modes disappear, because e.g. there's no external process that could disappear at any moment. On the other hand, it also means that bugs in the NVIDIA code, especially the Cgo parts, could take down the whole kubelet.

As for our case: we train deep learning models once, then run inference (recognise entities) on images, many times. The problem is that we have a lot of models and the number is increasing. They vary a lot in size, usually in proportion to the number of things they recognise. The general model is large, as it needs to recognise anything from dogs to people to houses, etc. We have smaller ones that detect only wedding stuff (brides, grooms, flowers, etc.: I'm not kidding) or whether content is safe for work. And we make smaller, custom ones for specific clients. Their traffic levels vary a lot, too. We need to run many replicas of the general model, but fewer of the others. In some cases, one replica is plenty. Even then, it's a waste to reserve a whole, expensive GPU to something that runs less than 1% of the time. Instead, we could fit (and already do so) several of those smaller models to run on the same device. We'd like to do the same under Kubernetes, by saying "pod X requires 512MB of GPU RAM", "pod Y requires 2GB". We'd use autoscaling to take care of the case when there's excessive contention among different models sharing the same GPU.

hongchaodeng commented on docs/proposals/gpu-support.md:
I see. The unit of resource sharing is one device, am I right? More specifically, CUDA device.

Previously I was wondering how runtime could isolate devices and cores for each container. IIUC, kubelet will be responsible for managing device sharing. It looks like docker nvidia doesn't support such mapping yet -- that plugin is just to make the nvidia driver agnostic. I think that's why we need to query the NVIDIA for device information. As a side note, this is quite different than how we are dealing with CPU: (not very sure) we are using container runtime API to create containers and let container runtime handle isolation. Underlying is cgroup doing isolation.

therc commented:
> Just to be clear, you're not suggesting we treat GPU processing and memory separately, right? Just that we treat GPU as fractional?

Good question. I don't think it's reasonable to expect perfect packing in both dimensions. Maybe we could add the constraint that the user may ask for only one or the other; the scheduler will then derive for them what fraction of the GPU to reserve. Otherwise, if an user is concerned mainly about GPU memory and we only let them express reservations in terms of fractions of GPUs, things will get messy once they have a mix of bare metal machines with 6GB and 12GB cards (like we do...). But this is not a problem yet for v0: all EC2 GPUs have 4GB of RAM, so for now we can just accept and manage fractional GPUs.

> Please include an explanation of how this works when you write the step-by-step instructions. It wasn't clear to me where the Go client library comes in (is it linked into Kubelet and used to start the GPU-using container?).

I'll add a clarification to the document. I meant that a Go client library is what Kubernetes already uses today in pkg/kubelet/dockertools to start Docker containers (we're migrating from github.com/fsouza/go-dockerclient to github.com/docker/engine-api). We don't build Docker command lines by hand, so nvidia-docker-plugin's way of returning command line flags is not optimal.

hongchaodeng commented:
> We'd use autoscaling to take care of the case when there's excessive contention among different models sharing the same GPU.

This is something I'm concerned about. First, how could you do isolation? I'm not worried about the programs but users abusing it. Second, how could you detect it?

therc commented:
If you don't want to share, you just reserve a whole GPU. There's not going
to be perfect isolation, but if others start to behave unfairly, you'll get
alerts (and if you're not monitoring and alerting, you're not running a
service, as someone said). You detect contention either indirectly, by
observing overall latency, or more directly, by measuring time spent inside
and outside the computation kernels. In order to compare against our SLOs,
we already gather other metrics to figure how much time is spent on various
operations while serving a request (e.g. fetching an image from an external
site when we're given only an URL), so that's easy.

See also what's done in the MPI world:
https://docs.nvidia.com/deploy/pdf/CUDA_Multi_Process_Service_Overview.pdf

hongchaodeng commented:
I totally agree with the autoscaling point.
I'm concerned about what happens if resource limit is exceeded? In such case we have to migrate some pods to another machines. Do you have anything in mind on this? Like, how to choose which pods to migrate to which machines?

I think the simplest form is that we could ignore it at the beginning (v0). Users could reserve more just in case.

hongchaodeng commented:
Of course, that's also a tough question as we are discuss in [DisruptionBudget](https://github.com/kubernetes/kubernetes/issues/12611)

therc commented:
In v0 we'll just hand out a whole device, so resources can't be exceeded. In v1, I think we'd use the same policies as for other resources. I like the idea of pod priorities, as in https://github.com/kubernetes/kubernetes/pull/22217

I just updated the document with the latest feedback: I added a first-class resource `nvidiaGpu`, got rid of the extender, added an example timeline of events and cleaned up things here and there. I haven't studied the scheduler in depth yet, so there are probably still errors and omissions here and there.

Hui-Zhi commented:
@therc so far the code of mine still need to be refined before submit it. After all the below issues fixed, I think it's ready for submitting. And I don't think they could take long time. Please let me know if you have any suggestion about them. You're an expert on NVIDIA GPU.
1. The code of nvidia-docker involve a dynamic library libnvidia-ml.so.1 for kubelet, it need modify the link arguments to compile it into kubelet binary.
2. Right now it need cuda environment to compile kubelet, so it need modify kubelet compile script to make the compile in a container.
3. Add a proper copyright into the new files, but the copyright may complex because I use the nvidia-docker project code and modify them.
4. Refine code and add meaningful comment and remove the debug code.
5. Add the unit-test code.
6.  Right now the implementation of gpu plugin supports multiple gpu computing platforms, but some code places use direct way to invoke cuda plugin function, it need modify them(I think only two or three places. Think about the place like gpuPlugin[0]).
7. Because the nvidia-docker use local path to import its project package, so I had to clone a new project and modify them to the web path and add them into kubelet.  It causes kubernetes project is hard to sync change with nvidia-docker project.  We need think about how to resolve it. Perhaps we could discuss with the maintainer of nvidia-docker to let them modify their import path.
8. I use the nvidia-docker way which use docker command to interact with docker daemon, but not docker-tools in cuda plugin.  It need replace them with docker-tools ways.
9. So far gpu plugins only suppport docker, but not rkt in kubenetes.  I don't test if use rkt what will happen(compile error? runtime error?).  However, right now the implementation of rkt in kubernetes don't support several interfaces which docker support, so I have to drop support the rkt, but it must to resolve it in the future.(Maybe the rkt maintainer of kubernetes can help)

hongchaodeng commented:
@yifan-gu can help answer the rkt question

timothysc commented on docs/proposals/gpu-support.md:
```diff
@@ -0,0 +1,339 @@
+## Objective
+
+Users should be able to request GPU resources for their workloads, as easily as
+for CPU or memory. Kubernetes should keep an inventory of machines with GPU
```
The resources are non-fungible, history has taught us there needs to be: 

1. GPU Card Count 
2. Card Version (May be better abstracted with newer versions of CUDA)
3. Driver+Library Version 

In a heterogeneous env. all three can vary and your software can depend on those three variables.

hongchaodeng replied:
Right. We should put it explicitly in the docs.
FYI, in Monday's meeting. We talked about the same thing.

dchen1107 replied:
@davidopp in another thread suggested to describe this kind non-fungible resource using a different approach called opaque counted resource at: https://github.com/kubernetes/kubernetes/issues/19049#issuecomment-170292193

I think that approach can cover the three attributes listed above by @timothysc. So that Kubelet can expose something like:

GPU-Version+Driver-Version: count through NodeCapacity / NodeAllocatable. 

Meanwhile, we also talked about this in today's sig-node meeting on the initial support (v0) with more restriction: 
* Only allow one GPU per node
* Introducing an experimental flag which includes GPU version info, and hardcode Kubelet to report GPU as NodeCapacity & NodeAllocatable if the flag is on. Also report this information through node label.
* Do we need to have a flag to config host device associating with GPU flag? I forgot our decision on this one.
* In this approach, required libraries have to be included in image itself. A given pod requesting a GPU, Kubelet based on the configuration to setup the runtime environment. 

cc/ @erictune Did I miss anything above? 

Once initContainer or VolumeContainer work is done, we can use that feature to allow preconfiged nvidia-docker volume driver, and Kubelet can discover those, and expose those using opaque counted resource approach above.

davidopp replied:
> Only allow one GPU per node

Can you explain the reason for this restriction? From the scheduler's perspective, it is no harder to deal with a resource with capacity 1 than a resource with capacity N.

> hardcode Kubelet to report GPU as NodeCapacity & NodeAllocatable if the flag is on.

How much code would you have to add to Kubelet to have Kubelet auto-detect the GPU(s) today, without init container or volume container?

timothysc replied:
@dchen1107 re: @davidopp 's  #19049 (comment) - totally agree. 

We used to call this arbitrary machine local limits, which other systems copied the basic design:  ( xref https://htcondor-wiki.cs.wisc.edu/index.cgi/tktview?tn=2905 ) 

Step 1: on startup there would be an option to kick a "plugin|discovery_script" whose output would be name value pairs of capabilities to be used for scheduling.    

Step 2: On allocate/match there would be another accompanying plugin/script to provision/assign said resource to a job, in this case would be pod and the machines status would be refreshed to ensure we decremented the resource count so the scheduler did the appropriate thing. 

In being generic you can apply arbitrary resource labeling.  (gpus/fpgas/infiniband cards/booster rockets/ion cannons/etc.) ;-)

timothysc commented on docs/proposals/gpu-support.md:
```diff
@@ -0,0 +1,339 @@
+- different vendors expose the hardware to users in different ways
+- some vendors require fairly tight coupling between the kernel driver
+controlling the GPU and the libraries/applications that access the hardware
+- it adds more resource types (whole GPUs, GPU cores, GPU memory)
```
I've never seen a cluster manager being able to isolate on anything other then card boundaries.

therc replied:
It doesn't qualify as a cluster manager, but our legacy environment packs together multiple processes on the same card through static allocation. At some point I'd like to do something similar with Kubernetes. We can't keep using the legacy pipeline because we're fed up with VMs. :-)  Even assuming exclusive use of the card, some might want to declare "I want to run this on a card with at least x GB". We have some workloads that fit the beefier hardware in our colocated machines, but not on AWS setups. They really only care about the memory.

timothysc replied:
Yeah, but iirc there is no memory protection, so an errant process could corrupt *.  In the case of multiple processes sharing the same card, you could simply leverage as a POD, but there should be no cross POD splicing of a GPU.   

We strictly need to prohibit that kind of abuse, b/c there is no memory protection.

therc replied:
Yeah, there are caveats, but it's done in MPI land: https://docs.nvidia.com/deploy/pdf/CUDA_Multi_Process_Service_Overview.pdf

I agree with you on disallowing sharing for arbitrary pods, but within the same namespace/user (and perhaps additional selectors, such as a common "build version" label), the user should be able to opt in to device sharing. You can easily aggregate multiple containers into the same pod if the makeup of the processes is consistent or predictable, but that's not always the case. Wouldn't pod affinity help here?

hongchaodeng replied:
> Wouldn't pod affinity help here?

I think let's not make it complicated at first step, i.e. coupling with unnecessary things.

I agree that we could try giving user the capability to opt in to device sharing. This sounds like real use case, although conservation of resources sharing should be done.

vishh commented:
Can we start with a user defined resource, `alpha.kubernetes.io/nvidia-gpu` for example, to expose this resource?

therc commented:
@vishh that sounds reasonable and gives it the "use at your own risk" feeling it needs. Would it do fitting checks in kubelet.canAdmitPod() or in predicates.podFitsResourcesInternal? I'm assuming the latter. I'm mostly worried about extra memory/CPU overhead for the 99.9% of users that won't use this.

Also, would it be worth adding to resource_helpers.go?

vishh commented:
@therc FYI: The feasibility check logic has changed recently. The [logic](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2284) is now shared with the scheduler.

What is the source of the extra memory overhead? 

Given that GPUs seem to need first class support in kubelet, and kubelet is not extensible to support new resources, **yes**, I imagine we will need a change to `resource_helpers.go` as well. As long as we reserve the right to change the API and behavior on this feature, until we are more comfortable with the overall system design, I guess we can punt on many of the issues for now.

Hui-Zhi commented:
@timothysc Are you asking for the below information discovered by kubelet? You can find it at:
https://github.com/Hui-Zhi/kubernetes/blob/try_nvidia_docker_merge_devices_support/pkg/kubelet/gpu/types/types.go

the code I have tested with 2 machines, and 1 machine installed 2 NVIDIA GM107GL [Quadro K2200] card. So far, if I start 2 containers in different pod, and each 1 need 1 GPU, then can be scheduled correctly. 
type GPUDevice struct {
        Path   string
        Cores  uint
        Memory uint
        Family string
}

type GPUPlatform struct {
        Name            string
        DriverVersion   string
        PlatformVersion string
}

type GPUDevices struct {
        GPUState
        GPUPlatform
        Devices []GPUDevice
}

Hui-Zhi commented:
@hongchaodeng could you let me know when the meeting start at UTC time? also the link to login? I also work on this. Thanks in advance.

hongchaodeng commented:
@therc @davidopp 
rkt update: It should work if you mount devices and libraries in rkt. We tried and successfully ran "nvidia-smi" test successfully.

therc commented:
@vishh thanks for the LGTM on the PR removing the dead function you linked to. :-)

For memory overhead, I was thinking of structs like [resourceRequest](https://github.com/kubernetes/kubernetes/blob/e93c0d727fb40b752411c3f7b9ab20c5d53ffecd/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L379). I'm assuming that's using 16 bytes now, but then I don't know how many of those build up between GC passes on a busy cluster.

Hui-Zhi commented:
@vishh @therc @davidopp  Regarding the resourceRequest and scheduler changes for GPU, could we just implement this just as https://github.com/kubernetes/kubernetes/pull/21446/files in the first version? 

I added the nvidia GPU resource in CheckPodsExceedingFreeResources

The CPU/Memory are 
fitsCPU := totalMilliCPU == 0 || (totalMilliCPU-milliCPURequested) >= podRequest.milliCPU
fitsMemory := totalMemory == 0 || (totalMemory-memoryRequested) >= podRequest.memory

But GPU only can be scheduled when it has resources, like:
fitsNvidiaGPU := (totalNvidiaGPU - nvidiaGPURequested) >= podRequest.nvidiaGPU

The Resource Type of GPU is:
Digit           = Format("Digit")

The resource_helpers.go should add:
+// Returns the NvidiaGPU number.
 +func (self *ResourceList) NvidiaGPU() *resource.Quantity {
 +	if val, ok := (*self)[ResourceNvidiaGPU]; ok {
 +		return &val
 +	}
 +	return &resource.Quantity{}
 +}
 +

Hui-Zhi commented:
@vishh regarding expose the nvidia GPU, I think maybe we only need to change the resource and scheduler as https://github.com/kubernetes/kubernetes/pull/21446/files
The implementation just like @davidopp said, "the minimum requirement for the first version is to treat GPU as a first-class integer resource"

I have tested the codes, and it could schedule GPUs, and pods hang when there don't have enough GPUs, the entire implementation you can find at:
https://github.com/Hui-Zhi/kubernetes/tree/try_nvidia_docker_merge_devices_support

The code of kubelet part also could work, but I need to refine it as you want.

The yaml for pod could be like:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nvidia-cuda
spec:
  containers:
  - name: nvidia-cuda
    image: nvidia_cuda:v1
    command: [ "sh", "/test.sh"]
    resources:
      requests:
        cpu: 100m
        memory: 100Mi
        nvidiagpu: 1
```

erictune commented:
Read proposal.  Didn't get a chance to read through all the conversations.  Proposal LGTM.
